### PR TITLE
Limit scope of CN name constraints

### PR DESCRIPTION
**Pull request by vdukhovni**
Don't apply DNS name constraints to the subject CN when there's a
least one DNS-ID subjectAlternativeName.

Don't apply DNS name constraints to subject CN's that are sufficiently
unlike DNS names.  Checked name must have at least two labels, with
all labels non-empty, no trailing '.' and all hyphens must be
internal in each label.  In addition to the usual LDH characters,
we also allow "_", since some sites use these for hostnames despite
all the standards.

##### Checklist

- [x ] tests are added or updated

### CODE REVIEW

**Comment by mattcaswell** [comment]
Are you planning to backport this? And if so to which branches?

File: crypto/x509v3/v3_ncons.c [hunk 1]
```diff
@@ -297,47 +297,143 @@ int NAME_CONSTRAINTS_check(X509 *x, NAME_CONSTRAINTS *nc)
...
     ASN1_STRING stmp;
     GENERAL_NAME gntmp;
+
     stmp.flags = 0;
     stmp.type = V_ASN1_IA5STRING;
     gntmp.type = GEN_DNS;
     gntmp.d.dNSName = &stmp;
 
+    gens = X509_get_ext_d2i(x, NID_subject_alt_name, NULL, NULL);
+    if (gens) {
```
Comment (mattcaswell) [inline crypto/x509v3/v3_ncons.c]: Should be `gens != NULL`

File: crypto/x509v3/v3_ncons.c [hunk 2]
```diff
@@ -297,47 +297,143 @@ int NAME_CONSTRAINTS_check(X509 *x, NAME_CONSTRAINTS *nc)
...
+        return utf8_length;
+    }
+    OPENSSL_free(utf8_value);
+    *dnsid = NULL;
+    return 0;
+}
+
+/*
+ * Check CN against DNS-ID name constraints, provided no DNS-ID
+ * subjectAlternativeName valies are present in the certificate.
```
Comment (mattcaswell) [inline crypto/x509v3/v3_ncons.c]: Typo: values

File: crypto/x509v3/v3_ncons.c [hunk 3]
```diff
@@ -297,47 +297,143 @@ int NAME_CONSTRAINTS_check(X509 *x, NAME_CONSTRAINTS *nc)
...
+
+            gen = sk_GENERAL_NAME_value(gens, i);
+            if (gen->type == GEN_DNS) {
+                dnsid_present = 1;
+                break;
+            }
+        }
+        GENERAL_NAMES_free(gens);
+    }
+    if (dnsid_present)
```
Comment (mattcaswell) [inline crypto/x509v3/v3_ncons.c]: Perhaps this check could be moved inside the above if (i.e. immediately after `GENERAL_NAMES_free()`. Also the declaration of `dnsid_present` could be moved inside the if.

File: crypto/x509v3/v3_ncons.c [hunk 4]
```diff
@@ -297,47 +297,143 @@ int NAME_CONSTRAINTS_check(X509 *x, NAME_CONSTRAINTS *nc)
 
 }
 
+#define CN2DNSID_NOMEM   (-2)
+#define CN2DNSID_BADNAME (-1)
+
+static int cn2dnsid(ASN1_STRING *cn, unsigned char **dnsid)
+{
+    int utf8_length;
```
Comment (mattcaswell) [inline crypto/x509v3/v3_ncons.c]: As this is a size of something it should probably be a size_t.

File: crypto/x509v3/v3_ncons.c [hunk 5]
```diff
@@ -297,47 +297,143 @@ int NAME_CONSTRAINTS_check(X509 *x, NAME_CONSTRAINTS *nc)
...
+     */
+    while (utf8_length > 0 && utf8_value[utf8_length-1] == '\0')
+        --utf8_length;
+
+    /* Reject *embedded* NULs */
+    if ((size_t)utf8_length != strlen((char *)utf8_value))
+        return CN2DNSID_BADNAME;
+
+    /*
+     * XXX: loose DNS name syntax, also check names with '_'
```
Comment (mattcaswell) [inline crypto/x509v3/v3_ncons.c]: What does the XXX mean here?

File: crypto/x509v3/v3_ncons.c [hunk 6]
```diff
@@ -297,47 +297,143 @@ int NAME_CONSTRAINTS_check(X509 *x, NAME_CONSTRAINTS *nc)
...
+     * Check DNS name syntax, any '-' or '.' must be internal,
+     * and on either side of each '.' we can't have a '-' or '.'.
+     *
+     * If the name has just one label, we don't consider it a DNS name.  This
+     * means that "CN=sometld" cannot be precluded by DNS name constraints, but
+     * that is not a problem.
```
Comment (mattcaswell) [inline crypto/x509v3/v3_ncons.c]: Why do we consider that not to be a problem?

File: crypto/x509v3/v3_ncons.c [hunk 7]
```diff
@@ -297,47 +297,143 @@ int NAME_CONSTRAINTS_check(X509 *x, NAME_CONSTRAINTS *nc)
...
+        if (i > 0 && i < utf8_length - 1) {
+            if (c == '-')
+                continue;
+            /*
+             * Next to a dot the preceding and following characters must not be
+             * another dot or a hyphen.  Otherwise, record that the name is
+             * plausible, since it has two or more labels.
+             */
+            if (c == '.'
+                && utf8_value[i-1] != '.' && utf8_value[i+1] != '.'
```
Comment (mattcaswell) [inline crypto/x509v3/v3_ncons.c]: Isn't it sufficient just to check that the preceding character wasn't '.'? If there's one after it we'll catch it next time around the loop.

File: crypto/x509v3/v3_ncons.c [hunk 8]
```diff
@@ -297,47 +297,143 @@ int NAME_CONSTRAINTS_check(X509 *x, NAME_CONSTRAINTS *nc)
...
+#define CN2DNSID_BADNAME (-1)
+
+static int cn2dnsid(ASN1_STRING *cn, unsigned char **dnsid)
+{
+    int utf8_length;
+    int i;
+    unsigned char *utf8_value;
+    int isdnsname = 0;
+
+    if ((utf8_length = ASN1_STRING_to_UTF8(&utf8_value, cn)) < 0)
```
Comment (mattcaswell) [inline crypto/x509v3/v3_ncons.c]: How does this work with internationalised domain names?

File: crypto/x509v3/v3_ncons.c [hunk 9]
```diff
@@ -297,47 +297,143 @@ int NAME_CONSTRAINTS_check(X509 *x, NAME_CONSTRAINTS *nc)
 
 }
 
+#define CN2DNSID_NOMEM   (-2)
+#define CN2DNSID_BADNAME (-1)
+
+static int cn2dnsid(ASN1_STRING *cn, unsigned char **dnsid)
```
Comment (mattcaswell) [inline crypto/x509v3/v3_ncons.c]: I'm not at all keen on overloading the return value of functions. This returns a length (which should really be a size_t) or error codes (which are negative), and tries to merge the two concepts into an int. This causes real problems with lengths/sizes of things being cast from one type to another and then back again. We should avoid that.

File: crypto/x509v3/v3_ncons.c (hunk 8 above)
Comment (mattcaswell) [reply crypto/x509v3/v3_ncons.c]: Why do we do this conversion to UTF8 here?

File: crypto/x509v3/v3_ncons.c (hunk 1 above)
Comment (vdukhovni) [reply crypto/x509v3/v3_ncons.c]: Fixing.

File: crypto/x509v3/v3_ncons.c (hunk 2 above)
Comment (vdukhovni) [reply crypto/x509v3/v3_ncons.c]: Fixing.

File: crypto/x509v3/v3_ncons.c (hunk 3 above)
Comment (vdukhovni) [reply crypto/x509v3/v3_ncons.c]: Fixing.

File: crypto/x509v3/v3_ncons.c (hunk 7 above)
Comment (vdukhovni) [reply crypto/x509v3/v3_ncons.c]: You're right, fixing, but checking the "next" rather than previous to catch any problem sooner.

File: crypto/x509v3/v3_ncons.c (hunk 8 above)
Comment (vdukhovni) [reply crypto/x509v3/v3_ncons.c]: Adding comments to answer these questions, the logic is correct and will remain the same.

File: crypto/x509v3/v3_ncons.c (hunk 9 above)
Comment (vdukhovni) [reply crypto/x509v3/v3_ncons.c]: I could separate out the return value from the length.  Note that given that the output is the result of a conversion via `ASN1_STRING_to_UTF8()` we're still stuck with a signed `int` length. If we ever support size_t lengths for ASN1_STRING types, then this code would be able to support that if refactored to separate the length from success/failure status.  Would like to see that done now, even though we don't yet get any benefit?

File: crypto/x509v3/v3_ncons.c [hunk 10]
```diff
@@ -323,7 +335,7 @@ static int cn2dnsid(ASN1_STRING *cn, unsigned char **dnsid)
         return CN2DNSID_BADNAME;
 
     /*
-     * XXX: loose DNS name syntax, also check names with '_'
+     * XXX: Deviation from strinct DNS name syntax, also check names with '_'
```
Comment (FdaSilvaYY) [inline crypto/x509v3/v3_ncons.c]: s/strinct/strict/

File: crypto/x509v3/v3_ncons.c [hunk 11]
```diff
@@ -350,8 +362,9 @@ static int cn2dnsid(ASN1_STRING *cn, unsigned char **dnsid)
              * plausible, since it has two or more labels.
              */
             if (c == '.'
-                && utf8_value[i-1] != '.' && utf8_value[i+1] != '.'
-                && utf8_value[i-1] != '-' && utf8_value[i+1] != '-') {
+                && utf8_value[i+1] != '.'
```
[the remark this answers is not in the document]
Comment (vdukhovni) [reply crypto/x509v3/v3_ncons.c]: Are you asking me to put spaces around the `+`.  I don't think this enhances the readability of array indices, for me the opposite is true.  I think of `foo[i+1]` as an atom in expression, and putting spaces around the `+` distracts from the larger structure of the containing expression.

File: crypto/x509v3/v3_ncons.c (hunk 10 above)
Comment (vdukhovni) [reply crypto/x509v3/v3_ncons.c]: Thanks.

File: crypto/x509v3/v3_ncons.c [hunk 12]
```diff
@@ -297,47 +297,152 @@ int NAME_CONSTRAINTS_check(X509 *x, NAME_CONSTRAINTS *nc)
...
+    int utf8_length;    /* Return type of ASN1_STRING_to_UTF8 */
+    int i;
+    unsigned char *utf8_value;
+    int isdnsname = 0;
+
+    /*
+     * Per RFC 6125 DNS-IDs representing internationalized domain names appear
+     * in certificates in A-label encoded form.  The same applies to CNs which
+     * are intended to represent DNS names.  However, while in the SAN DNS-IDs
+     * are IA5Strings as CNs they may be needlessly stored in 16-bit Unicode.
```
Comment (dot-asm) [inline crypto/x509v3/v3_ncons.c]: Aren't some commas missing? I mean it seems unnecessary hard to parse...

File: crypto/x509v3/v3_ncons.c [hunk 13]
```diff
@@ -297,47 +297,152 @@ int NAME_CONSTRAINTS_check(X509 *x, NAME_CONSTRAINTS *nc)
...
+     * We perform a conversion to UTF-8 here, to ensure that we get an ASCII
+     * representation of strings that have been encoded in some other way in
```
Comment (dot-asm) [inline crypto/x509v3/v3_ncons.c]: "ASCII representation" is misleading in the context. I mean ASCII implies 7-bit chars, and ASCII representation of Unicode is UTF-7, not UTF-8. Well, one can always say "you know what I mean", and everybody kind of does, so this would be a side note... Otherwise one can probably use "'char' representation" (when 'char' refers to C type). or "C string representation"...

File: crypto/x509v3/v3_ncons.c [hunk 14]
```diff
@@ -297,47 +297,152 @@ int NAME_CONSTRAINTS_check(X509 *x, NAME_CONSTRAINTS *nc)
...
+     * that is not a problem.
+     */
+    for (i = 0; i < utf8_length; ++i) {
+        unsigned char c = utf8_value[i];
+
+        if ((c >= 'a' && c <= 'z')
+            || (c >= 'A' && c <= 'Z')
+            || (c >= '0' && c <= '9')
+            || c == '_')
+            continue;
```
Comment (dot-asm) [inline crypto/x509v3/v3_ncons.c]: There is kind of ambiguity between CN conversion to UTF8 and this check. I mean if non-ASCII characters are permitted to appear in CN, and host does have non-ASCII in its name, then would it be reasonable to expect that it's "human-friendly" one? Of course RFC says that subjectAltName is the recommended one, CN-based is kind of for backward compatibility. And one can as well argue that non-ASCII chars in hostname are not *that* backward compatible...

File: crypto/x509v3/v3_ncons.c (hunk 9 above)
Comment (mattcaswell) [reply crypto/x509v3/v3_ncons.c]: Personally I would prefer the length to be separated. I do think we should be looking to sort out the size_t/int mess at some point (we pretty much have in libssl, except where it touches libcrypto). So I'd rather not dig the hole any deeper. However, it's not in our coding style, so I won't insist on it.

**Comment by vdukhovni** [comment]
I think this is ready.  Any final review comments?

File: crypto/x509v3/v3_ncons.c [hunk 15]
```diff
@@ -297,48 +297,155 @@ int NAME_CONSTRAINTS_check(X509 *x, NAME_CONSTRAINTS *nc)
...
+     */
+    if ((utf8_length = ASN1_STRING_to_UTF8(&utf8_value, cn)) < 0)
+        return X509_V_ERR_OUT_OF_MEM;
+
+    /*
+     * Some certificates have had names that include a *trailing* NUL byte.
+     * Remove these harmless NUL characters. They would otherwise yield false
+     * alarms with the following embedded NUL check.
+     */
+    while (utf8_length > 0 && utf8_value[utf8_length-1] == '\0')
```
Comment (dot-asm) [inline crypto/x509v3/v3_ncons.c]: Styling nit even here, `utf8_length - 1`.

**Comment by vdukhovni** [comment]
I pushed a final fixup for the missing spaces Andy noticed around another binary minus in an index expression.  I'm assuming this does not change approval so I'll merge later today, but in the meantime will open a squashed backport for 1.1.0 and 1.0.2 (IIRC also affected).

**Comment by vdukhovni** [comment]
So the backport is just to 1.1.0. #6295 unless we decide that we should add (the newly llmited) subject CN name constraint checks in 1.0.2 for consistency (Rich is not keen on this).

Please review the 1.1.0 delta (identical modulo merge conflict resolution).

**Comment by vdukhovni** [comment]
I decided to check the documentation to see whether we have any text about name constraints that might revision. DId find that but did find that we have two flags that affect the behaviour of X509_check_host(3) in ways relevant to this PR.

* X509_CHECK_FLAG_ALWAYS_CHECK_SUBJECT
* X509_CHECK_FLAG_NEVER_CHECK_SUBJECT

The PR should probably be updated to take these into account.  Ignoring the presence of SANs when the "always" flag is set, and checking the CN regardless, or never checking the subject CN when the "never" flag is set.

However, there is a potential minor drawback. The verification state of the certificate is saved as part of hte client side session, if a session is reused, but the flags are different, they have no effect, since the certificate checks happened in an earlier connection. That's probably just fine, the client should not do session resumption with a different policy than the original session. So I think we ignore this rather exotic edge-case as unsupported client behaviour.  [ For example, if the client sets a different reference identifier (hostname to check) it has no effect with session resumption, the original chain will be reported authenticated. ]

**Comment by vdukhovni** [comment]
Anyone?  I am include to include a commit to take the flags into account...

**Comment by t-j-h** [comment]
A commit to handle those flags as described makes sense to me.

**Comment by mattcaswell** [comment]
> Anyone? I am include to include a commit to take the flags into account...

Yes, I think you should.

**Comment by vdukhovni** [comment]
New commit pushed, please review.

File: crypto/x509v3/v3_ncons.c [hunk 16]
```diff
@@ -297,48 +297,140 @@ int NAME_CONSTRAINTS_check(X509 *x, NAME_CONSTRAINTS *nc)
...
+    if ((utf8_length = ASN1_STRING_to_UTF8(&utf8_value, cn)) < 0)
+        return X509_V_ERR_OUT_OF_MEM;
+
+    /*
+     * Some certificates have had names that include a *trailing* NUL byte.
+     * Remove these harmless NUL characters. They would otherwise yield false
+     * alarms with the following embedded NUL check.
+     */
+    while (utf8_length > 0 && utf8_value[utf8_length - 1] == '\0')
+        --utf8_length;
```
Comment (t-j-h) [inline crypto/x509v3/v3_ncons.c]: Style-wise I wouldn't --utf8_length as pre-increment isn't being used ...

File: crypto/x509/x509_vfy.c [hunk 17]
```diff
@@ -658,7 +679,12 @@ static int check_name_constraints(X509_STORE_CTX *ctx)
                 int rv = NAME_CONSTRAINTS_check(x, nc);
 
                 /* If EE certificate check commonName too */
-                if (rv == X509_V_OK && i == 0)
+                if (rv == X509_V_OK && i == 0
+                    && !(ctx->param->hostflags
+                         & X509_CHECK_FLAG_NEVER_CHECK_SUBJECT)
```
Comment (mattcaswell) [inline crypto/x509/x509_vfy.c]: Style nit: should really be "== 0"

File: crypto/x509/x509_vfy.c [hunk 18]
```diff
@@ -658,7 +679,12 @@ static int check_name_constraints(X509_STORE_CTX *ctx)
...
+                    && ((ctx->param->hostflags
+                         & X509_CHECK_FLAG_ALWAYS_CHECK_SUBJECT)
```
Comment (mattcaswell) [inline crypto/x509/x509_vfy.c]: Style nit: should really be "!= 0"

File: doc/man3/X509_VERIFY_PARAM_set_flags.pod [hunk 19]
```diff
@@ -133,14 +133,29 @@ B<name> clearing any previously specified host name or names.  If
...
 are strongly advised to use this interface in preference to explicitly
-calling L<X509_check_host(3)>, hostname checks are out of scope
+calling L<X509_check_host(3)>, hostname checks may be out of scope
 with the DANE-EE(3) certificate usage, and the internal check will
-be suppressed as appropriate when DANE support is added to OpenSSL.
+be suppressed as appropriate when DANE verification is enabled.
+
+When the subject CommonName will not be ignored, whether as a result of the
+B<X509_CHECK_FLAG_ALWAYS_CHECK_SUBJECT> host flag, or because no DNS subject
+alternaitve names are present in the certificate, any DNS name constraints in
```
Comment (mattcaswell) [inline doc/man3/X509_VERIFY_PARAM_set_flags.pod]: Typo: alternative

**Comment by vdukhovni** [comment]
@mattcaswell Matt, are you still a reviewer for the final version, or shall I just proceed with Tim?

**Comment by mattcaswell** [comment]
> Matt, are you still a reviewer for the final version, or shall I just proceed with Tim?

Well, I approved subject to my nits being fixed above.

**Comment by vdukhovni** [comment]
> @mattcaswell Well, I approved subject to my nits being fixed above.

Thanks, will fix nits and push.
